File: lab1/1.js
const a0 = 16387;
const betta = 16387;
const K = 48;
const M = 2 ** 31;
const epsilon = 0.05;

// Quite bad solution, but lgtm
const generatorPrototype = Object.getPrototypeOf(function* () { }()).__proto__;
generatorPrototype.getMultipleValues = function (n) {
    return new Array(n).fill().map(() => this.next().value);
}

function* linearCongruentialGenerator(a, betta, c, M) {
    while (true) {
        a = (a * betta + c) % M;
        yield a;
    }
}

function* multiplexialCongruentialGenerator(a, betta, M) {
    yield* linearCongruentialGenerator(a, betta, 0, M);
}

function* macLarenMarsagliaGenerator(XGenerator, YGenerator, MY, k) {
    const V = XGenerator.getMultipleValues(k);

    while (true) {
        const [X, Y] = [XGenerator.next().value, YGenerator.next().value];
        const j = Math.floor(k * Y / MY);
        yield V[j];
        V[j] = X;
    }
}

/**
 * Magic function from https://ru.wikipedia.org/wiki/%D0%9A%D0%B2%D0%B0%D0%BD%D1%82%D0%B8%D0%BB%D0%B8_%D1%80%D0%B0%D1%81%D0%BF%D1%80%D0%B5%D0%B4%D0%B5%D0%BB%D0%B5%D0%BD%D0%B8%D1%8F_%D1%85%D0%B8-%D0%BA%D0%B2%D0%B0%D0%B4%D1%80%D0%B0%D1%82
 */
function approximateHiSquared(alpha, n) {
    const d = alpha < 0.5 ?
        -2.0637 * (Math.log(1 / alpha) - 0.16) ** 0.4274 + 1.5774 :
        2.0637 * (Math.log(1 / (1 - alpha)) - 0.16) ** 0.4274 - 1.5774;
    const a = [
        1.0000886,
        0.4713941,
        0.000134802,
        -0.008553069,
        0.00312558,
        -0.0008426812,
        0.00009780499
    ];
    const b = [
        -0.2237368,
        0.02607083,
        0.01128186,
        -0.01153761,
        0.005169654,
        0.00253001,
        -0.001450117,
    ];
    const c = [
        -0.01513904,
        -0.008986007,
        0.02277679,
        -0.01323293,
        -0.006950356,
        0.001060438,
        0.001565326,
    ];
    return n * (a.reduce((ac, el, i) => ac + n ** (-i / 2) * d ** i * (
        el + b[i] / n + c[i] / n ** 2
    ), 0)) ** 3;
}

function hiSquaredTest(values, M, epsilon, k = 20) {
    const nu = new Array(k).fill(0);
    values.forEach(el => nu[Math.floor(el * k / M)]++);
    const p_k = values.length / k;
    const hiSquared = nu.reduce((sum, nu_k) => sum + ((nu_k - p_k) ** 2) / p_k, 0);
    const delta = approximateHiSquared(1 - epsilon, k - 1);
    if (hiSquared < delta) {
        return { nu, str: `Passed hi-squared test: ${hiSquared.toFixed(5)} < ${delta.toFixed(5)}` };
    } else {
        return { nu, str: `Hi-squared test failed: ${hiSquared.toFixed(5)} > ${delta.toFixed(5)}` };
    }
}

function approximateKolmogorov(epsilon) {
    let res = 0;
    for (let i = -1000; i <= 1000; i++) {
        res += (-1) ** i * Math.exp(-2 * k ** 2 * epsilon ** 2)
    }
    return res;
}

function kolmogorovTest(values, M, epsilon) {
    // To avoid sorting of original array
    let copiedValues = [...values];
    copiedValues.sort((a, b) => a - b);
    const Dn = copiedValues.reduce((ac, val, i) => Math.max(ac, Math.abs(val / M - (i + 1) / copiedValues.length)), 0) *
        Math.sqrt(copiedValues.length);
    // sadly, but I took it from table :c
    const Kolmagorov_Quantile = 1.36;
    if (Dn < Kolmagorov_Quantile) {
        return `Passed Kolmogorov Test: ${Dn.toFixed(2)} < ${Kolmagorov_Quantile.toFixed(2)}`;
    } else {
        return `Kolmagorov test failed: ${Dn.toFixed(5)} > ${Kolmagorov_Quantile.toFixed(5)}`;
    }
}

const multiplexialCongruentialResult = multiplexialCongruentialGenerator(a0, betta, M).getMultipleValues(1000);
const macLarenMarsagliaResult = macLarenMarsagliaGenerator(
    linearCongruentialGenerator(a0, betta, 0, M),
    linearCongruentialGenerator(a0 + 42, betta - 42, 42, M),
    M, K
).getMultipleValues(1000);


const { nu: macLarenMarsagliaNu, str: macLarenMarsagliaHiStr } = hiSquaredTest(macLarenMarsagliaResult, M, epsilon, 10);
const { nu: multiplexialCongruentialNu, str: multiplexialCongruentialHiStr } = hiSquaredTest(multiplexialCongruentialResult, M, epsilon, 10);
const multiplexialCongruentialKolmagorovStr = kolmogorovTest(multiplexialCongruentialResult, M, epsilon);
const macLarenMarsagliaKolmagorovStr = kolmogorovTest(macLarenMarsagliaResult, M, epsilon);
/**
 * DEPRECATED
 */
// function approximateHiSquared(alpha, n) {
//     const d = alpha < 0.5 ?
//         -2.0637 * (Math.log(1 / alpha) - 0.16) ** 0.4274 + 1.5774 :
//         2.0637 * (Math.log(1 / (1 - alpha)) - 0.16) ** 0.4274 - 1.5774;
//     const E = d * (9 * d ** 4 + 256 * d ** 2 - 433) / (4860 * Math.SQRT2);
//     const D = - (6 * d ** 4 + 14 * d ** 2 - 32) / 405;
//     const C = d * (d ** 2 - 7) / (9 * Math.SQRT2);
//     const B = 2 / 3 * (d ** 2 - 1);
//     const A = d * Math.SQRT2;
//     const sqrtN = Math.sqrt(n);
//     return n + A * sqrtN + B + C / sqrtN + D / n + E / (n * sqrtN);
// }

File: lab1/1.java
import java.util.Arrays;

public class M {
  static double[] MCG(long a, long b, long M, int n) {
     double[] res = new double[n];
     for(int i = 0; i < n; i++) {
        a = a * b % M;
        res[i] = a * 1. / M;
     }
     return res;
  }
  
  static double[] MLMG(double[] X, double[] Y, int k, int n) {
      double[] V = new double[k];
      for(int i = 0; i < k; i++) {
          V[i] = X[i];
      }
      double[] res = new double[n];
      for(int i = 0; i < n; i++) {
          int j = (int)Math.floor(Y[i] * k);
          res[i] = V[j];
          V[j] = X[i+k];
      }
      return res;
  }
  
  static boolean hiSquared(double[] X) {
      int[] nu = new int[10];
      for(int i = 0; i < X.length; i++) {
          nu[(int)Math.floor(10 * X[i])]++;
      }
      double p_k = X.length / 10.;
      double hiSquared = 0;
      for(int i = 0; i < 10; i++) {
           hiSquared += Math.pow(nu[i] - p_k, 2) / p_k;
      }
      System.out.println(hiSquared);
      return hiSquared < 16.919;
  }
  
  static boolean KT(double[] X) {
      Arrays.sort(X);
      double Dn = 0;
      for (int i = 0; i < X.length; i++) {
          Dn = Math.max(Dn, Math.abs(X[i] - (i + 1) * 1. / X.length));
      }
      Dn *= Math.sqrt(X.length);
      System.out.println(Dn);
      return Dn < 1.36;
  }
  
  public static void main(String[] args) {
    double[] mcl = MCG(68921, 68921, 2147483648L, 1000);
    double[] mlmg = MLMG(MCG(68921, 68921, 2147483648L, 1048), MCG(68921, 68921, 2147483648L, 1000), 48, 1000);
    System.out.println(hiSquared(mcl));
    System.out.println(hiSquared(mlmg));
    System.out.println(KT(mcl));
    System.out.println(KT(mlmg));
  }
};